DirectProgramming/C++SYCL_FPGA/Tutorials/DesignPatterns/memory_system/README.md

```diff
@@ -0,0 +1,256 @@
+
+In the `NaiveKernel`, the first array dimension (row) is unrolled in a loop. In this way, the kernel tries to access all five elements in one column of the buffer, namely `buffer[0][col]` to `buffer[4][col]`. As a result, there are 4 loads and 5 stores to the memory system in each loop iteration, and since the initiation interval is 1, they happens in a single clocl cycle. The memory layout for the `buffer` in this case has 8 banks. Looking at the Details from the Kernel Memory Viewer, we can learn that the bank bits are second to the fourth least significant bits $b_4$, $b_3$, $b_2$. This essentially means that integers in every other column will be in a different memory bank.
+
+|                | Address | Bank bits | Bank # |
+|----------------|---------|-----------|--------|
+| buffer[0][0]   | 0x0000  | 0b000     | 0      |
+| buffer[0][1]   | 0x0004  | 0b001     | 1      |
+| buffer[0][2]   | 0x0008  | 0b010     | 2      |
+| buffer[0][3]   | 0x000C  | 0b011     | 3      |
+| buffer[0][4]   | 0x0010  | 0b100     | 4      |
+| ...            |         |           |        |
+| buffer[0][499] | 0x07CC  | 0b011     | 3      |
+| buffer[1][0]   | 0x07D0  | 0b100     | 4      |
+| ...            |         |           |        |
+
+We can produce this table showing which integer is stored in which bank of the memory system. 
+
+![Naive Kernel Memory Layout](./assets/memory_system_naive.png)
+
+Observe that each access to one column of the `buffer` creates 2 stores and 3 loads to even numbered memory banks and 2 stores and 2 loads to odd numbered memory banks. These number of load and store surpasses the physical capacity of the memory blocks, and therefore, the compiler inferrs stallable LSUs as arbitration node in the Kernel Memory Viewer.
+
+![Stallable Arbitration Node](./assets/memory_report_naive.png)
+
+The stallable arbitration inside the loop body will stall the downstream logic from making progress, and in this example, we expect stall in pipe write for the `OutStream_NaiveKernel` pipe of which it's output depends on the current column of memory that is being modified. We can verify this behaviour in the simulation waveform: there are constant "dips" in the valid signal for pipe write - i.e. the valid signal is desserted because of the upstream stall from the memory system. This can negatively affect the system throughput and cause outer loop to have serial execution.
+
+![Naive System Waveform](./assets/waveform_naive.png)
+
+### Optimized Kernel
+
+To make the memory system more efficient, we can make three changes:
+
+1. We use the right-most array dimension for unrolled accesses, i.e. column dimension.
+
+2. We choose the number of banks to a power-of-2 using `constexpr`-math functions. This line of code computes the closest power-of-2 above the given number. We use this power-of-2 as the number of banks in our optimized memory system. In our case, we use 8.
```

Comment:
```suggestion
2. Choose the number of banks to a power-of-2 using `constexpr`-math functions. This line of code computes the closest power-of-2 above the given number. We use this power-of-2 as the number of banks in our optimized memory system. Since this design performs up to 5 parallel memory accesses, a minimum of 5 banks are required. Since the next power-of-2 is 8, this design should be specified with 8 banks.
```